### DISCUSSION TITLE
Add socket activation support by preloading libsdsock with LD_PRELOAD

### DISCUSSION THREAD

**Discussion by eriksjolund** [containers/podman#23546] [Show and tell]
Status: experimental

Is it possible to use _socket activation_ even when the executable in the container image does not support _socket activation_?

Yes, sometimes.
If the executable in the container is dynamically linked, it might be possible to preload the library [__libsdsock__](https://github.com/ryancdotorg/libsdsock)
to add support for _socket activation_.

The container unit needs to set the environment variables `LD_PRELOAD` and `LIBSDSOCK_MAP`

```
Environment=LD_PRELOAD=/usr/local/lib/libsdsock.so
Environment=LIBSDSOCK_MAP=tcp://0.0.0.0:8080=myweb.socket
```

Also modify the container image so that it has the systemd libraries (RPM package: _systemd-libs_) and the file _/usr/local/lib/libsdsock.so_ installed.

I documented it as a step-by-step example

[_example: pasta + custom network + socket activation + libsdsock - source address preserved_](https://github.com/eriksjolund/podman-networking-docs?tab=readme-ov-file#example-pasta--custom-network--socket-activation--libsdsock---source-address-preserved)

**Comment by afbjorklund** [comment]
You could also use `systemd-socket-proxy`, for the good ole `inetd` feeling

https://www.freedesktop.org/software/systemd/man/latest/systemd-socket-proxyd.html

**Comment by eriksjolund** [comment]
True, but if you use `systemd-socket-proxy`, then the network server will not see the correct source IP address (i.e. the IP address of the client).

I think the main advantage of using __libsdsock__ with `LD_PRELOAD` is that the source address is preserved. There might also be a performance advantage compared to using `PublishPort` but I haven't done any benchmarks. The traffic sent over the socket-activated socket is sent directly to the network server (not via pasta).

Currently pasta does not support preserving the source IP address when a custom network is used. 

For example a quadlet setup like this does not support preserving the source address:

file: _~/.config/containers/systemd/mynet.network_
```
[Network]
Internal=true
```
file: _~/.config/containers/systemd/myweb.container_
```
[Container]
Image=localhost/myweb
Network=mynet.network
PublishPort=8080:8080

[Install]
WantedBy=default.target
```

Changing such a setup to use  __libsdsock__ with `LD_PRELOAD` instead would fix the source address problem.
